orttraining/orttraining/test/python/orttraining_test_hierarchical_ortmodule.py

```diff
@@ -0,0 +1,141 @@
+#!/usr/bin/env python3
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+import torch
+import torch.fx
+import torch.nn as nn
+import torch.nn.functional as F
+from torch.utils.checkpoint import checkpoint
+from onnxruntime.training.ortmodule import ORTModule
+from onnxruntime.training.ortmodule.experimental.hierarchical_ortmodule import HierarchicalORTModule
+
+
+class A(nn.Module):
+    # A supported module.
+    def __init__(self):
+        super(A, self).__init__()
+        self.l1 = nn.Linear(2, 2)
+
+    def forward(self, x):
+        return self.l1(x)
+
+
+class B(nn.Module):
+    # This module is not exportable to ONNX because it
+    # uses gradient-checkpointing. However, its two sub-module's
+    # are exportable, so ORTModule should be used to compute them.
+    def __init__(self):
+        super(B, self).__init__()
+        self.l1 = nn.Linear(2, 2)
+        self.a = A()
+
+    def forward(self, x):
+        def custom():
+            def custom_forward(x_):
+                return self.a(x_)
+            return custom_forward
+        z = self.l1(checkpoint(custom(), x))
+        return z
+
+
+class C(nn.Module):
+    # A supported module.
+    def __init__(self):
+        super(C, self).__init__()
+        self.l1 = nn.Linear(2, 2)
+
+    def forward(self, x):
+        y = F.relu(self.l1(x))
+        return y
+
+
+class D(nn.Module):
+    # This module is not exportable to ONNX because it
+    # inner module self.b uses gradient-checkpointing.
+    def __init__(self):
+        super(D, self).__init__()
+        self.b = B()
+
+    def forward(self, x):
+        y = F.sigmoid(self.b(x))
+        return y
+
+
+class Main(nn.Module):
```

Comment:
can we add cases to have non-tensor inputs, multiple module outputs, same value used as multiple outputs, etc...

Reply:
I feel this case doesn't need those tests. This classes just make a forward call to underlying nn.Modules and ORTModules, which is much simpler. This PR doesn't touch operator schema and those kind of things.